pkg/admissioncontroller/webhooks/admission/auditpolicy/admission.go

```diff
@@ -0,0 +1,294 @@
+	if !ok {
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failure to cast to audit Policy type: %v", schemaVersion))
+	}
+	errList := auditvalidation.ValidatePolicy(auditPolicyInternal)
+	if len(errList) != 0 {
+		return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("provided invalid audit policy err=%v", errList))
+	}
+
+	// Validate audit policy schema version against k8s version of shoot
+	if isValidVersion, err := IsValidAuditPolicyVersion(shoot.Spec.Kubernetes.Version, schemaVersion); err != nil {
+		return admission.Errored(http.StatusUnprocessableEntity, err)
+	} else if !isValidVersion {
+		err := fmt.Errorf("your shoot cluster version %q is not compatible with audit policy version %q", shoot.Spec.Kubernetes.Version, schemaVersion.GroupVersion().String())
+		return admission.Errored(http.StatusUnprocessableEntity, err)
+	}
+
+	return acadmission.Allowed("referenced audit policy is valid")
+}
+
+func (h *handler) admitConfigMap(ctx context.Context, request admission.Request) admission.Response {
+	var (
+		oldCm = &corev1.ConfigMap{}
+		cm    = &corev1.ConfigMap{}
+	)
+	if err := h.getOldObject(request, oldCm); err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+
+	if request.Operation == admissionv1.Delete {
+		if controllerutil.ContainsFinalizer(oldCm, shoot.FinalizerName) {
+			return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("deletion of audit-policy is forbidden as it is still referenced by a shoot (cm has finalizer)"))
+		}
+		return acadmission.Allowed("deletion of audit-policy accepted")
+	}
+
+	if request.Operation == admissionv1.Update {
+		if err := h.decoder.Decode(request, cm); err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		if !controllerutil.ContainsFinalizer(cm, shoot.FinalizerName) {
+			return acadmission.Allowed("configmap is not referenced by a Shoot")
+		}
+
+		auditPolicy, err := getAuditPolicy(cm)
+		if err != nil {
+			return admission.Errored(http.StatusUnprocessableEntity, err)
+		}
+
+		oldAuditPolicy, ok := oldCm.Data[auditPolicyConfigMapDataKey]
+		if ok {
+			if oldAuditPolicy == auditPolicy {
+				return acadmission.Allowed("audit policy did not change")
+			}
+		}
+
+		// Validate audit policy semantically
+		auditPolicyObj, schemaVersion, err := decoder.Decode([]byte(auditPolicy), nil, nil)
+		if err != nil {
+			return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("failed to decode the provided audit policy err=%v", err))
+		}
+		auditPolicyInternal, ok := auditPolicyObj.(*audit_internal.Policy)
+		if !ok {
+			return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failure to cast to audit Policy type: %v", schemaVersion))
+		}
+		errList := auditvalidation.ValidatePolicy(auditPolicyInternal)
+		if len(errList) != 0 {
+			return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("provided invalid audit policy err=%v", errList))
+		}
```

Comment:
I haven't checked in detail, but this is the same code like above, or? Maybe you can move it inside a function and call it from both places?

Reply:
I tried to do that but gave up after about 20minutes. The problem is that we need the `schemaVersion` later and then we need at least 3 return values (schema version, error Code and error message) for a function. Then this function is kind of unintuitive validating but also returning a version -  so I left it like it is.

Reply:
Hm, I see, it wouldn't improve it too much. OK then.